lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -1645,28 +1648,23 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
             builder->create<rtl::ConstantOp>(APInt(log2(readLatency + 1), j));
         auto enJ = builder->create<sv::ArrayIndexInOutOp>(enReg, jIdx);
         auto addrJ = builder->create<sv::ArrayIndexInOutOp>(addrReg, jIdx);
-        readPipe.push_back({enJ, addrJ});
+        auto rdEnJ = builder->create<sv::ReadInOutOp>(enJ);
+        auto rdAddrJ = builder->create<sv::ReadInOutOp>(addrJ);
+        readPipe.push_back({enJ, addrJ, rdEnJ, rdAddrJ});
       }
       if (readLatency != 0) {
         addToAlwaysFFBlock(EventControl::AtPosEdge, clk, [&]() {
-          auto lastEn = readPipe[0].en;
-          auto lastAddr = readPipe[0].addr;
           for (size_t j = 1; j < readLatency + 1; ++j) {
-            auto thisEn = readPipe[j].en;
-            auto thisAddr = readPipe[j].addr;
-
-            auto readLastEn = builder->create<sv::ReadInOutOp>(lastEn);
-            builder->create<sv::PAssignOp>(thisEn, readLastEn);
-            addIfProceduralBlock(readLastEn, [&]() {
-              builder->create<sv::PAssignOp>(
-                  thisAddr, builder->create<sv::ReadInOutOp>(lastAddr));
+            builder->create<sv::PAssignOp>(readPipe[j].en,
+                                           readPipe[j - 1].rd_en);
+            addIfProceduralBlock(readPipe[j - 1].rd_en, [&]() {
+              builder->create<sv::PAssignOp>(readPipe[j].addr,
+                                             readPipe[j - 1].rd_addr);
```

Comment:
Just using the indices here is cleaner than the original fold-like implementation. Nice.